src/model/markercollection.js

```diff
@@ -0,0 +1,287 @@
+ * For licensing, see LICENSE.md.
+ */
+
+import LiveRange from './liverange';
+import Position from './position';
+import Range from './range';
+import EmitterMixin from '@ckeditor/ckeditor5-utils/src/emittermixin';
+import CKEditorError from '@ckeditor/ckeditor5-utils/src/ckeditorerror';
+import mix from '@ckeditor/ckeditor5-utils/src/mix';
+
+/**
+ * Creates, stores and manages {@link ~Marker markers}.
+ *
+ * Markers are created by {@link ~MarkerCollection#set setting} a name for a {@link module:engine/model/liverange~LiveRange live range}
+ * in `MarkerCollection`. Name is used to group and identify markers. Names have to be unique, but markers can be grouped by
+ * using common prefixes, separated with `:`, for example: `user:john` or `search:3`.
+ *
+ * Since markers are based on {@link module:engine/model/liverange~LiveRange live ranges}, for efficiency reasons, it's
+ * best to create and keep at least markers as possible.
+ */
+export default class MarkerCollection {
+	/**
+	 * Creates a markers collection.
+	 */
+	constructor() {
+		/**
+		 * Stores {@link ~Marker markers} added to the collection.
+		 *
+		 * @private
+		 * @member {Map} #_markers
+		 */
+		this._markers = new Map();
+	}
+
+	/**
+	 * Returns an iterator that iterates over all {@link ~Marker markers} added to the collection.
+	 *
+	 * @returns {Iterator}
+	 */
+	[ Symbol.iterator ]() {
+		return this._markers.values();
+	}
+
+	/**
+	 * Checks if marker with given `markerName` is in the collection.
+	 *
+	 * @param {String} markerName Marker name.
+	 * @returns {Boolean} `true` if marker with given `markerName` is in the collection, `false` otherwise.
+	 */
+	has( markerName ) {
+		return this._markers.has( markerName );
+	}
+
+	/**
+	 * Returns {@link ~Marker marker} with given `markerName`.
+	 *
+	 * @param {String} markerName Name of marker to get.
+	 * @returns {~Marker|null} Marker with given name or `null` if such marker was not added to the collection.
+	 */
+	get( markerName ) {
+		return this._markers.get( markerName ) || null;
+	}
+
+	/**
+	 * Creates and adds a {@link ~Marker marker} to the `MarkerCollection` with given name on given
+	 * {@link module:engine/model/range~Range range}.
+	 *
+	 * If `MarkerCollection` already had a marker with given name (or {@link ~Marker marker} was passed) and the range to
+	 * set is different, the marker in collection is removed and then new marker is added. If the range was same, nothing
+	 * happens and `false` is returned.
+	 *
+	 * @fires {module:engine/model/markercollection~MarkerCollection#event:add}
+	 * @fires {module:engine/model/markercollection~MarkerCollection#event:remove}
+	 * @param {String|~Marker} markerOrName Name of marker to add or Marker instance to update.
+	 * @param {module:engine/model/range~Range} range Marker range.
+	 * @returns {~Marker} `Marker` instance added to the collection.
+	 */
+	set( markerOrName, range ) {
```

Comment:
Only name, why would it take `range`? You need to pass `range` to update the marker. Of course we can keep it simple because there's not much more to for developer if we restrict this only to marker name. It's just `.set( marker.name, range )` instead `.set( marker, range )`.